### PR TITLE
NIFI-3290 Reporting task to send bulletins with S2S

### PR DESCRIPTION
**Pull request by pvillard31**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [X] Is your initial contribution a single, squashed commit?

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [X] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by olegz** [comment]
Reviewing. . .

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        final List<PropertyDescriptor> properties = new ArrayList<>();
+        properties.add(DESTINATION_URL);
+        properties.add(PORT_NAME);
+        properties.add(SSL_CONTEXT);
+        properties.add(COMPRESS);
+        properties.add(TIMEOUT);
+        properties.add(PLATFORM);
+        return properties;
+    }
```
Comment (olegz) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: @pvillard31 given the fact that this operation is invoked multiple times consider applying the same pattern we use in Processors where property descriptor list (as well as relationships) are build in static initializers.

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+                arrayBuilder.add(serialize(factory, builder, bulletin, df, platform, nodeId));
+            }
+        }
+        final JsonArray jsonArray = arrayBuilder.build();
+
+        // Send the JSON document for the current batch
+        try {
+            final Transaction transaction = getClient().createTransaction(TransferDirection.SEND);
+            if (transaction == null) {
+                getLogger().debug("All destination nodes are penalized; will attempt to send data later");
```
Comment (olegz) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: Wondering if this should be INFO

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                    new Object[]{bulletins.size(), transferMillis, transactionId, bulletins.get(0).getId()});
+        } catch (final IOException e) {
+            throw new ProcessException("Failed to send Bulletins to destination due to IOException:" + e.getMessage(), e);
+        }
+
+        // Store the id of the last event so we know where we left off
+        try {
+            StateManager stateManager = context.getStateManager();
+            Map<String, String> newMapOfState = new HashMap<>();
+            newMapOfState.put(LAST_EVENT_ID_KEY, String.valueOf(currMaxId));
```
Comment (olegz) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: More of a style issue, but I wonder if we should use ```Collections.singletonMap(key, value)``` or do you expect /know already that there may be more values?

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        }
+
+        lastSentBulletinId = currMaxId;
+    }
+
+    private Long getMaxBulletinId(List<Bulletin> bulletins) {
+        long result = -1L;
+        for (Bulletin bulletin : bulletins) {
+            if(bulletin.getId() > result) {
+                result = bulletin.getId();
```
Comment (olegz) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: A bit confused. What if several _bulletin.getId()_ are greater then _result_ are you expecting the last one? And if so is there some predefined ordering?
Alternatively I am thinking that you may want to break out of the loop once _result_ is assigned, right?

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java (hunk 1 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: Yes you're right!

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java (hunk 2 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: That's fair, I changed the level.

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java (hunk 3 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: I don't imagine other values at the moment so I changed to a singleton map. That's more concise.

File: nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java (hunk 4 above)
Comment (pvillard31) [inline nifi-nar-bundles/nifi-site-to-site-reporting-bundle/nifi-site-to-site-reporting-task/src/main/java/org/apache/nifi/reporting/SiteToSiteBulletinReportingTask.java]: Hmmm, just trying to get the maximum ID from a list of bulletins. I do need to parse the whole list to get the maximum ID (unless I assume there is an order). I updated the PR with a more concise approach. Let me know if I missed something.

**Pull request closed without merging**